session: add webrequest api

deepak1556 opened the pull request:
```
session.webRequest.<method>([filter,] function(details) {
  return result
})
```
where result and details will be same as `chrome.webRequest.details` and [chrome.webRequest.BlockingResponse](https://developer.chrome.com/extensions/webRequest#type-BlockingResponse)

- [x] Filter with URL regex

Fixes #2115

paulcbetts commented:
Remember when I said I'm not super excited about this? I'm actually *very* excited about this, because it lets you use Chrome's networking stack to make requests from the main process - this is important because if you use node.js's networking stack (i.e. `request` and friends), it doesn't respect system proxy settings / NTLM auth / etc etc etc.

deepak1556 commented:
@zcbenz basic functionality is done. Left out stuff would be adding the ability to modify form data before sending request. Also have left out auth part of the request lifecycle since there is a login event to handle the stuff, should it still be repeated here ?

steelbrain commented:
:heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart:

christoth commented:
+ 1

zcbenz commented on atom/browser/api/atom_api_web_request.cc:
```diff
@@ -0,0 +1,87 @@
+void WebRequest::SetListener(mate::Arguments* args) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+
+  base::DictionaryValue* filter = new base::DictionaryValue();
```
Using raw pointer is very easy to leak (and it does here), we should wrap it in a `scoped_ptr`.

zcbenz commented on atom/browser/api/atom_api_web_request.cc:
```diff
@@ -0,0 +1,87 @@
+
+// static
+void WebRequest::BuildPrototype(v8::Isolate* isolate,
+                             v8::Local<v8::ObjectTemplate> prototype) {
```
Unaligned indent.

zcbenz commented on atom/browser/api/atom_api_web_request.h:
```diff
@@ -0,0 +1,48 @@
+
+#include "atom/browser/api/trackable_object.h"
+#include "atom/browser/net/atom_network_delegate.h"
+#include "base/callback.h"
```
Seems redundant.

zcbenz commented on atom/browser/api/atom_api_web_request.h:
```diff
@@ -0,0 +1,48 @@
+class WebRequest : public mate::TrackableObject<WebRequest> {
+ public:
+  static mate::Handle<WebRequest> Create(v8::Isolate* isolate,
+                                      AtomBrowserContext* browser_context);
```
Unaligned indent.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,377 @@
+  return true;
+}
+
+base::DictionaryValue* ExtractRequestInfo(net::URLRequest* request) {
```
Should return `scoped_ptr<base::DictionaryValue>` or `base::DictionaryValue`, this pointer is leaked.

zcbenz commented on atom/browser/net/atom_network_delegate.h:
```diff
@@ -0,0 +1,104 @@
+ public:
+  struct BlockingResponse;
+  using Listener =
+      base::Callback<BlockingResponse(const base::DictionaryValue*)>;
```
It should accept `const base::DictionaryValue&`.

zcbenz commented on atom/common/native_mate_converters/value_converter.h:
```diff
@@ -24,6 +24,12 @@ struct Converter<base::DictionaryValue> {
 };
 
 template<>
+struct Converter<const base::DictionaryValue*> {
```
I don't think we need to add a converter for the pointer type, we can just do something like `Emit('blabla', *dict)` in code.

deepak1556 commented:
@zcbenz have addressed the comments.

zcbenz commented on atom/browser/api/atom_api_web_request.cc:
```diff
@@ -0,0 +1,87 @@
+}
+
+template<AtomNetworkDelegate::EventTypes type>
+void WebRequest::SetListener(mate::Arguments* args) {
```
We need a way to clear previous listener, we can make `SetListenerInIO` clear the listen when `callback` is `null`.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+std::string ResourceTypeToString(content::ResourceType type) {
+  switch (type) {
+    case content::RESOURCE_TYPE_MAIN_FRAME:
+      return "main_frame";
```
Should be `mainFrame`.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+    case content::RESOURCE_TYPE_OBJECT:
+      return "object";
+    case content::RESOURCE_TYPE_XHR:
+      return "xmlhttprequest";
```
`xhr` is a well-known abbreviation.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+  return dict.Pass();
+}
+
+base::DictionaryValue* GetRequestHeadersDict(
```
The `base::DictionaryValue*` should also be wrapped in `scoped_ptr` here,  and in `GetResponseHeadersDict` bellow.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+  return header_dict;
+}
+
+base::DictionaryValue* GetResponseHeadersDict(
```
There is a converter for `net::HttpResponseHeaders` in `api::WebContents`, it should be reused.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+
+// static
+std::map<AtomNetworkDelegate::EventTypes, AtomNetworkDelegate::ListenerInfo>
+AtomNetworkDelegate::event_listener_map_;
```
It should be a member data, a multi-session app can have multiple `NetworkDelegate`s.

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+    net::URLRequest* request,
+    const net::CompletionCallback& callback,
+    GURL* new_url) {
+  brightray::NetworkDelegate::OnBeforeURLRequest(request, callback, new_url);
```
We should check whether `kOnBeforeRequest` is in `event_listener_map_`, and if it is then execute code bellow, otherwise execute super class's method.

All overrides bellow should do the same.

deepak1556 replied:
we have `ignore-connection-limit` logic setup in `brightray::OnBeforeURLRequest` , should that be moved here if we are going to call super class method optionally ?

zcbenz commented on atom/browser/net/atom_network_delegate.cc:
```diff
@@ -0,0 +1,388 @@
+    auto details = ExtractRequestInfo(request);
+
+    BrowserThread::PostTaskAndReplyWithResult(BrowserThread::UI, FROM_HERE,
+        base::Bind(&RunListener, wrapped_callback, base::Passed(&details)),
```
Does `base::Bind(wrapped_callback, *details.release())` work?

deepak1556 replied:
Nope, compiier reports `error: field of type 'base::DictionaryValue' has private copy constructor`

zcbenz commented on atom/browser/net/atom_network_delegate.h:
```diff
@@ -0,0 +1,104 @@
+  };
+
+  struct ListenerInfo {
+    ListenerInfo() {}
```
Constructor and destructor are not needed.

zcbenz commented on atom/browser/net/atom_network_delegate.h:
```diff
@@ -0,0 +1,104 @@
+  };
+
+  struct BlockingResponse {
+    BlockingResponse() {}
```
Should initialize `cancel` in the constructor.

zcbenz commented on atom/browser/net/atom_network_delegate.h:
```diff
@@ -0,0 +1,104 @@
+    BlockingResponse() {}
+    ~BlockingResponse() {}
+
+    int Cancel() const {
```
`Cancel()` usually means an action, can probably just use `code()`.

zcbenz commented on atom/browser/net/atom_network_delegate.h:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    bool cancel;
+    GURL redirectURL;
```
C++ code should use `redirect_url `, the same rule also applies for `requestHeaders` and `responseHeaders`.

zcbenz commented on docs/api/session.md:
```diff
@@ -286,3 +286,178 @@ myWindow.webContents.session.setCertificateVerifyProc(function(hostname, cert, c
+    * `timestamp` Double
+    * `redirectURL` String
+    * `statusCode` Integer
+    * `ip` String - The server IP address that the request was actually sent to.
```
This is optional too.

deepak1556 commented:
@zcbenz have addressed the comments, except for one question. Also `defaultSession` returned by `session` module wasnt persistent, have fixed that too.

zcbenz commented:
:+1:
